### PR TITLE
Assignment 5 - Zach Lawson and Tom Rohr

### PR DESCRIPTION
**Pull request by zlawson-ewu**
Assignment 5 peer programmed by Zach Lawson (@zlawson-ewu) and Tom Rohr (@Tom-Rohr), no EC attempted.

### CODE REVIEW

**Comment by joshmerrell** [comment]
I will review this.

**Comment by joshmerrell** [comment]
It looks pretty good overall, and has very high test coverage.

## Instructions

- Create a *class library* project called "GenericsHomework.". ✔
- Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
  - Define the `Node` class  ✔
  - Include a constructor that takes a value.  (No validation is necessary on the value). ✔
  - Add a `ToString()` override that writes out the value's `ToString()` result. ✔
  - Add a `Next` property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
    - The `Next` property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
    - The `Next` property setter should be private. ✔
  - Add an `Append` method that takes a value and appends a new `Node` instance after the current node (by invoking the `Next` property). ✔
  - Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
    - Whether it is sufficient to only set Next to itself ✔
    - Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
    - Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
  - Create an Exists method to test to see if a value exists in the list. ✔
  - Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
- You should not rely on any BCL generic classes for your implementation. ✔

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

1. Implement a `VennDiagram` structure that contains `n` `Circle`s that only contains homogenous **reference types** of any type. ❌

- Each circle contains n items and each item can belong to one or more `Circle` instances.
- You are not required to use a `Node` from earlier in the homework for your Venn diagram implementation.
- You are welcome to use existing BCL generic classes for the extra credit.

1. Implement `System.Collections.Generic.ICollection<T>` on the `Node` class ❌

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.  ✔
- Place all shared project items into a `Directory.Build.targets` file.  ❌
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

File: GenericsHomework/GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    public void Append(T value)
+    {
+        if (!Exists(value))
+        {
+            Node<T> temp = new(value) { Next = Next };
+            Next = temp;
+        }
+        else
+        {
+            throw new ArgumentException(nameof(value) + " already exists in the list.");
```
Comment (joshmerrell) [inline GenericsHomework/GenericsHomework/Node.cs:24]: Consider using an interpolated string here.

Comment (zlawson-ewu) [reply GenericsHomework/GenericsHomework/Node.cs:24]: A fair point. I did it this way out of habit really.

File: GenericsHomework/GenericsHomework.Test/NodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
+namespace GenericsHomework.Test;
+
+[TestClass]
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Test/NodeTests.cs:3]: I would recommend adding a few more tests for testing your .Next property, you can check the assignment solution for some examples.

File: GenericsHomework/GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
+﻿namespace GenericsHomework;
+
+public class Node<T>
+{
+    public T? Value { get; set; } //Assuming "No validation is necessary on the value" means null is allowed.
```
[the remark this answers is not in the document]
Comment (zlawson-ewu) [reply GenericsHomework/GenericsHomework/Node.cs:5]: Ah I see, thank you for clearing that up.

File: GenericsHomework/GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            }
+            temp = temp.Next;
+        } while (!temp.Equals(this) && !exists);
+        return exists;
+    }
+
+    public override string? ToString()
+    {
+        return (Value is null) ? "null" : Value!.ToString();
+    }
```
[the remark this answers is not in the document]
Comment (zlawson-ewu) [reply GenericsHomework/GenericsHomework/Node.cs:46]: Agreed, I intended to make this change after some clarification in Thursday's class, but it slipped my mind.

File: GenericsHomework/GenericsHomework.Test/NodeTests.cs (hunk 2 above)
Comment (zlawson-ewu) [reply GenericsHomework/GenericsHomework.Test/NodeTests.cs:3]: Ah I see, we should've tried more list sizes. Thank you for the tip.

**Pull request closed without merging**